### PR TITLE
docs: Add database design

### PR DESCRIPTION
**Pull request by dave-tucker**
Updates: #348 

Signed-off-by: Dave Tucker <dave@dtucker.co.uk>

### CODE REVIEW

File: docs/design/database.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
+Database Schema
+===============
```
Comment (russellb) [inline docs/design/database.md:2]: I think markdown lint will complain about this. It doesn't like mixing heading styles (underlining vs. the `#` prefixes)

File: docs/design/database.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+1. User runs the Apex agent
+1. Apex agent does a `POST /device` to register the unique device with Apex for the user. It is assigned into the single zone that a user is in (either the default zone, or a new zone where the user was moved to by an admin).
+1. Once the device is registered a `POST /zones/$id/peers` was issued to create an association between this device and a zone.
+1. `GET /zones/$id/peers` was polled to retrieve the peer listing for a zone.
+
+On top of this flow, we were somewhat abusing `POST /zones/$id/peers` to also update peers after NAT-T.
+
+This model was built under the assumption that multi-tenancy would be achieved by deploying an Apex stack per-tenant.
+
+# v2: SaaS (Milestone-1)
```
Comment (russellb) [inline docs/design/database.md:69]:
```suggestion
# v2: SaaS Preview Milestone
```

File: docs/design/database.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+In order to create a more SaaS-ready schema, we've made the following changes.
+
+1. Replace the `Zone` construct with `Organization`. This is a GitHub like construct, where each unique `User` is also it's own `Organization`.
```
Comment (russellb) [inline docs/design/database.md:73]:
```suggestion
1. Replace the `Zone` construct with `Organization`. This is a GitHub-like construct, where each unique `User` is also its own `Organization`.
```

File: docs/design/database.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+- Users: Unique users of the service
+- Devices: A unique end-user device
+- Organization: A collection of users who have permissions to onboard one or more devices into this organization
+
+Since a `User` is also an `Organization`, once someone has registered to Apex we onboard devices into their personal organization.
+
+The device-onboarding flow is as follows:
+
+1. A User registers for Apex using an sign-up form on https://apex-hosted.cloud, with the option to use SSO from Google, Github, Facebook etc..
```
Comment (russellb) [inline docs/design/database.md:120]:
```suggestion
1. A User registers for Apex using a sign-up form on <https://apex-hosted.cloud>, with the option to use SSO from Google, Github, Facebook etc..
```

markdownlint will complain about raw URLs.

File: docs/design/database.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+1. User downloads and runs the Apex agent
+1. Apex agent does a `POST /device` to register the unique device with Apex for the user - it's onboarded into their personal organization.
+1. `GET /organization/$id/devices` is polled to retrieve the peer listing for a zone.
+1. An dedicated API endpoint was added to update device information if the public-facing address changes etc...
+
+A key point here is that there is no admin intervention required to onboard a new user as anyone can sign-up and register.
+
+Supporting a user belonging to multiple organizations is easy in this model since a `Device` may only be in one `Organization` at a time. A `User` may chose to move their device at any time.
```
Comment (russellb) [inline docs/design/database.md:128]: ... but actually supporting this isn't expected at first, right?  You could call the "multi-organization" support v2.5 or something?

File: docs/design/database.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+
+# v3: (TBD) Milestone-2
+
+In our v3 schema we seek to add a few constructs like `Group` and `Network` to allow for more fine-grained connection policies.
```
Comment (russellb) [inline docs/design/database.md:133]: I don't want to bikeshed `Group` and `Network` name choices. How about just listing objectives we expect to go after next in future iterations?

```
In future versions of the schema, we seek to evolve the schema to support some additional key features:
* Allowing finer-grained connectivity grouping for devices within the same organization. We may want to allow subsets of devices with an organization to have their own mesh network.
* Allowing for security-policy for traffic within an Apex managed network.
```

If we have existing issues in the backlog for those features, it'd be good to cross-reference them here, too.

**Review by russellb — CHANGES_REQUESTED** [review]
Thanks for writing this up. I am generally supportive of the v2 iteration as described here. My comments are relatively minor.

File: docs/design/database.md (hunk 5 above)
Comment (dave-tucker) [reply docs/design/database.md:128]: I've added a note about this in the text that the functionality won't be implemented in v1->v2 but that the schema supports such a feature. It wouldn't require changes so we wouldn't be at v2.5.

File: docs/design/database.md (hunk 1 above)
Comment (dave-tucker) [reply docs/design/database.md:2]: Ack. Fixing now.

File: docs/design/database.md (hunk 6 above)
Comment (dave-tucker) [reply docs/design/database.md:133]: I removed this entirely. We can open an issue and another PR to track/discuss v3 changes.

File: docs/design/database.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+1. User runs the Apex agent
+1. Apex agent does a `POST /device` to register the unique device with Apex for the user. It is assigned into the single zone that a user is in (either the default zone, or a new zone where the user was moved to by an admin).
+1. Once the device is registered a `POST /zones/$id/peers` was issued to create an association between this device and a zone.
+1. `GET /zones/$id/peers` was polled to retrieve the peer listing for a zone.
+
+On top of this flow, we were somewhat abusing `POST /zones/$id/peers` to also update peers after NAT-T.
+
+This model was built under the assumption that multi-tenancy would be achieved by deploying an Apex stack per-tenant.
+
+## v2: SaaS Preview Milestone
```
Comment (russellb) [inline docs/design/database.md:68]:
```suggestion
## v2: SaaS Preview Milestone

See [issue #348](https://github.com/redhat-et/apex/issues/348).
```

**Review by russellb — APPROVED** [review]
lgtm.

My only other suggestion is to include a link to the issue tracking the v2 work so someone could go find the code associated with it if they were coming from this doc.

**Comment by russellb** [comment]
I'll hold off on merging in case anyone else wants to weigh in on this.

**Comment by nerdalert** [comment]
LGTM, next step to brainstorm is the slicing of the org. VPCification. 🎉

**Review by nerdalert — APPROVED** [review]
LGTM, next step to brainstorm is the slicing of the org. VPCification. 🎉

File: docs/design/database.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        string cidr
+        string hub_zone
+    }
+```
+
+In this simplified model we only have 3 concepts:
+
+- Users: Unique users of the service
+- Devices: A unique end-user device
+- Organization: A collection of users who have permissions to onboard one or more devices into this organization
```
Comment (vishnoianil) [inline docs/design/database.md:115]: I think we need some clarity around users <-> organization relationship. Looks like we support both the models N:1 and 1:N for user and organization. 1:N relation is clear. But needs some clarification around N:1 model. 

- Every user will have it's own personal organization, and given that each organization can have multiple users, I assume that the original owner of the organization need to invite other users to his organization? Will that invited user have similar kind of ownership on the invited organization (e.g can this invited user invite other users)? or we will have a concept of original owner and invitees, where invitee only have "read" only access? I assume it's the later, but want to clarify.
- In v2 we are not allowing users to join multiple organization, but do we allow user to join any other organization (1:1) ?

Comment (dave-tucker) [reply docs/design/database.md:115]: The diagram shows that a user may belong to 1 or more organizations - the first organization they belong to is their personal org.

> Every user will have it's own personal organization, and given that each organization can have multiple users, I assume that the original owner of the organization need to invite other users to his organization? Will that invited user have similar kind of ownership on the invited organization (e.g can this invited user invite other users)? or we will have a concept of original owner and invitees, where invitee only have "read" only access? I assume it's the later, but want to clarify.

The invite mechanism hasn't been designed yet - and is somewhat tied in with the ACL work we need to do across the API... but your assumption is correct. We'd have different permission levels in an org and the default would be read-only for everyone other than the owner.

> In v2 we are not allowing users to join multiple organization, but do we allow user to join any other organization (1:1) ?

I think this is backwards. V1 was 1:1 User:Zone. V2 is 1:Many(1 or more) User:Organization.
V2 can support joining multiple orgs, even if it takes a while for us to implement it. The main thing is the database layout wouldn't need to change

Comment (vishnoianil) [reply docs/design/database.md:115]: Sounds good.

**Review by vishnoianil — APPROVED** [review]

**Pull request merged**